osu.Game/Rulesets/RulesetStore.cs

```diff
@@ -48,7 +57,18 @@ public RulesetStore(IDatabaseContextFactory factory)
         /// </summary>
         public IEnumerable<RulesetInfo> AvailableRulesets { get; private set; }
 
-        private Assembly resolveRulesetAssembly(object sender, ResolveEventArgs args) => loadedAssemblies.Keys.FirstOrDefault(a => a.FullName == args.Name);
+        private Assembly resolveRulesetDependencyAssembly(object sender, ResolveEventArgs args)
+        {
+            var asm = new AssemblyName(args.Name);
+
+            // the requesting assembly may be located out of the executable's base directory, thus requiring manual resolving of its dependencies.
+            // this assumes the only explicit dependency of the ruleset is the game core assembly.
+            // the ruleset dependency on the game core assembly requires manual resolving, transitive dependencies should be resolved automatically
+            if (asm.Name.Equals(typeof(OsuGame).Assembly.GetName().Name, StringComparison.Ordinal))
```

Comment:
Loading the tau ruleset (https://github.com/Altenhh/tau) via the `rulesets/` directory fails on this branch. Loading via placing it alongside all other game executables works.
Seems like the reason is it's expecting an exact match for transitive dependencies, which will almost never be the case (frequent o!f updates).

Perhaps we don't want to have this automatic transitive dependency matching, and instead look for all local assemblies as an override to ruleset-provided ones, via:

```csharp
foreach (var curAsm in AppDomain.CurrentDomain.GetAssemblies())
{
    if (asm.Name.Equals(curAsm.GetName().Name, StringComparison.Ordinal))
        return curAsm;
}
```

Reply:
That's exactly it, yes. But it'll fail the other way around too - with tau referencing an earlier version of o!f and osu! referencing a more up-to-date one.

Reply:
The weird thing is that I also tested with another pre-built ruleset release ([maimai](https://github.com/LumpBloom7/osu-maimai) 2020.413.0) and it didn't fail, even though [it technically references an older version of framework than the game itself did](https://github.com/LumpBloom7/osu-maimai/blob/a5b221ba4867d9261b2a403686647a56879a8bb5/osu.Game.Rulesets.Maimai/osu.Game.Rulesets.Maimai.csproj#L21)... Probably needs more investigation/thought either way.

Reply:
The code path was assuming that the only explicit dependency was on the game core and so the resolved framework assembly version was the same as the game core but it isn't actually the case at all (looks like I took a shortcut in my thinking) as you pointed out , as rulesets can depend on a older framework version than the game depends on; while your proposed code path works fine (tested this with recent releases of tau, bosu, maimai and mvis rulesets),  there should be version checks for game version / framework version  added on top to handle ABI breaking changes, else game will crash when using the rulesets as there is a ABI mismatch as shown here:

![image](https://user-images.githubusercontent.com/20256717/79744447-a042db80-8306-11ea-90a8-c5208d4f3853.png)

Reply:
Ouch. I think that's fine, but it should stop the load of gameplay and throw a notification instead of crashing the game though. Can be done in a further PR.

Exception handling should probably exist everywhere where a ruleset is used, I'd say.

Reply:
In that case , will open a PR soon to mitigate this.

Reply:
Actually, this may not be as trivial as I thought. Fixing this may require a quite a bit of investment in addings guards where `DrawableRuleset`s are manipulated as well as a bit of thinking since some pieces of the `DrawableRuleset` aren't created directly in `Ruleset.CreateDrawableRulesetWith()` but lazily created (namely `Playfield`), where errors could pop up without the ability to catch them.